gsy_myco_sdk/matching_algorithms/preferred_partners_algorithm.py

```diff
@@ -179,3 +125,99 @@ def _get_required_energy_and_rate_from_order(
         else:
             order_required_clearing_rate = order["energy_rate"]
         return order_required_energy, order_required_clearing_rate
+
+    @classmethod
+    def _calculate_bid_offer_matches_for_one_market_timeslot(
+            cls, market_id: str, time_slot: str, data: Dict) -> List[BidOfferMatch]:
+        """
+        Calculate all possible matches for one market slot.
+        """
+        bid_offer_matches = []
+        bids = data.get("bids")
+        offers = data.get("offers")
+        # Sorted bids in descending orders
+        sorted_bids = sort_list_of_dicts_by_attribute(bids, "energy_rate", True)
+        # Sorted offers in descending order
+        sorted_offers = sort_list_of_dicts_by_attribute(offers, "energy_rate", True)
+        offers_mapping = cls._get_actor_to_offers_mapping(offers)
+        available_order_energy = {}
+        for bid in sorted_bids:
+            for offer in sorted_offers:
+                if offer.get("seller") == bid.get("buyer"):
+                    continue
+
+                possible_match = cls._match_one_bid_one_offer(
+                    offer, bid, available_order_energy, market_id, time_slot, offers_mapping)
+                if possible_match:
+                    bid_offer_matches.append(possible_match)
+
+                if (bid["id"] in available_order_energy and
+                        available_order_energy[bid["id"]] <= FLOATING_POINT_TOLERANCE):
+                    break
```

Comment:
unless the current bid's energy is totally used, keep matching this bid with more offers